File: Proteomic_Data_Analysis_Tsirakou.R
# The code below aims to perform proteomics data analysis (with artificial data), 
# pre-process them, statistically analyze them, and visualize the results.  
# The code was adapted from ChatGPT.
# For further information, you may visit https://openai.com/chatgpt


# First, we install the required packages and load the respective libraries
install.packages("ggplot2")
install.packages("pheatmap")
install.packages("dplyr")
install.packages("tidyr")
install.packages("stats")
install.packages("ggrepel")

library(ggplot2)
library(pheatmap)
library(dplyr)
library(tidyr)
library(stats)
library(ggrepel)

# Then, we set a random seed for reproducibility
set.seed(40)

# Parameters of our artificial/synthetic data
number_proteins <- 300
number_samples_per_group <- 10

# Simulating the control and treatment groups
control_group <- matrix(rnorm(number_proteins * number_samples_per_group, mean = 10, sd = 1),
                        nrow = number_proteins)

treatment_group <- matrix(rnorm(number_proteins * number_samples_per_group, mean = 10, sd = 1),
                          nrow = number_proteins)

# Let's say that 18 proteins are differentially expressed
differential_expressed_proteins <- sample(1:number_proteins, 18, replace = FALSE)
treatment_group[differential_expressed_proteins, ] <- treatment_group[differential_expressed_proteins, ] + matrix(rnorm(18 * number_samples_per_group, mean = 3, sd = 2,
                                                                          nrow = 18)

# Combining the data
data_combined <- cbind(control_group, treatment_group)
colnames(data_combined) <- c(paste0("Control_Group", 1:number_samples_per_group),
                             paste0("Treatment_Group", 1:number_samples_per_group))
rownames(data_combined) <- paste0("Protein_", 1:number_proteins)

# Pre-processing the data (log2 transformation)

data_log2 <- log2(data_combined + 1)

# Z-score normalization
data_z <- t(apply(data_log2, 1, scale))
rownames(data_z) <- rownames(data_log2)

# Principal Component Analysis

PCA_results <- prcomp(t(data_log2), scale. = TRUE)
PCA_data_frame <- data.frame(PCA_results$x[, 1:2])
PCA_data_frame$Group <- factor(c(rep("Control", number_samples_per_group), rep("Treatment", number_samples_per_group)))

# Plotting the Principal Components
ggplot(PCA_data_frame, aes(x = PC1, y = PC2, color = Group)) +
  geom_point(size = 4) +
  theme_minimal() +
  labs(title = "Principal Components of Proteomic Data")

# Generating the heatmap of the 18 differentially expressed proteins

# What is the variance?
protein_variance <- apply(data_log2, 1, var)
top18_differential_expression <- names(sort(protein_variance, decreasing = TRUE))[1:18]

# Plotting the heatmap
pheatmap(data_log2[top18_differential_expression, ],
         scale = "row",
         cluster_rows = TRUE,
         cluster_cols = TRUE,
         main = "Top 18 Most Variable Proteins",
         show_rownames = FALSE)

# Testing for differential expression 

group_labels <- c(rep("Control", number_samples_per_group), rep("Treatment", number_samples_per_group))

log2fc <- numeric(number_proteins)
pvals <- numeric(number_proteins)

for (i in 1:number_proteins) {
  control_vals <- data_log2[i, group_labels == "Control"]
  treatment_vals <- data_log2[i, group_labels == "Treatment"]
  
  t_test <- t.test(treatment_vals, control_vals)
  pvals[i] <- t_test$p.value
  log2fc[i] <- mean(treatment_vals) - mean(control_vals)
}

# Creating a data frame for the results
results <- data.frame(
  Protein = rownames(data_log2),
  log2FC = log2fc,
  p_value = pvals
)

# Adjusting the p-values (Benjamini-Hochberg)
results$adj_p <- p.adjust(results$p_value, method = "BH")

# Generating a volcano plot

results$Significant <- results$adj_p < 0.05

ggplot(results, aes(x = log2FC, y = -log10(p_value), color = Significant)) +
  geom_point() +
  scale_color_manual(values = c("green", "yellow")) +
  theme_minimal() +
  labs(title = "Volcano Plot", x = "Log2 Fold Change", y = "-Log10 p-value") +
  geom_vline(xintercept = c(-1, 1), linetype = "dashed", color = "darkgreen") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "blue")

File: Proteomic_Data_Analysis_Tsirakou.m
% The code below aims to perform proteomics data analysis (with artificial data), 
% pre-process them, statistically analyze them, and visualize the results.  
% The code was adapted from ChatGPT.
% For further information, you may visit https://openai.com/chatgpt


% First, we set a random seed for reproducibility
rng(50); % let's say 50

% Parameters for our artificial/synthetic proteomic data
number_proteins = 300;
number_samples_per_group = 40;

% Data
control_group = normrnd(40, 2, [number_proteins, number_samples_per_group]);
treatment_group = normrnd(40, 2, [number_proteins, number_samples_per_group]);

% Differential expression of 15 proteins
differential_expression_idx = randperm(number_proteins, 15);
treatment_group(differential_expression_idx, :) = treatment_group(differential_expression_idx, :) + normrnd(3, 0.5, [15 number_samples_per_group]);

% Combinining the data
data_combined = [control_group, treatment_group];
log2_data = log2(data_combined + 1);

% Pre=processing - Z-score normalization
data_z_score = (log2_data - mean(log2_data, 2)) ./ std(log2_data, 0, 2);

% Principal Component Analysis
[coeff, score, ~] = pca(log2_data');

% Plotting Principal Component Analysis
figure;
gscatter(score(:,1), score(:,2), [repmat({'Control'}, number_samples_per_group, 1); repmat({'Treatment'}, number_samples_per_group,1)]);
title('Principal Components');
xlabel('Component 1');
ylabel('Component 2');
grid on;

% Generating the heatmap of the 15 most variable proteins
variances = var(log2_data, 0, 2);
[~, idx] = maxk(variances, 15);
figure;
heatmap(log2_data(idx, :));
title('Top 15 Most Variable Proteins');

% Performing differential expression analysis
log2fc = mean(log2_data(:, number_samples_per_group+1:end), 2) - mean(log2_data(:, 1:number_samples_per_group), 2);
pvals = zeros(number_proteins, 1);

for i = 1:number_proteins
    [~, pvals(i)] = ttest2(log2_data(i, 1:number_samples_per_group), log2_data(i, number_samples_per_group+1:end));
end

% Adjusting p-values using Benjamini-Hochberg
[sorted_pvals, sort_idx] = sort(pvals);
adj_pvals = sorted_pvals .* number_proteins ./ (1:number_proteins)';
adj_pvals(adj_pvals > 1) = 1;
adj_pvals_original = zeros(size(adj_pvals));
adj_pvals_original(sort_idx) = adj_pvals;

% Generating a volcano plot
figure;
scatter(log2fc, -log10(pvals), 40, adj_pvals_original < 0.05, 'filled');
colormap([0 1 0; 1 1 0]); 
hold on;
yline(-log10(0.05), '--b');
xline(1, '--g');
xline(-1, '--g');
title('Volcano Plot');
xlabel('Log2 Fold Change');
ylabel('-Log10 p-value');
grid on;